cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternals.qll

```diff
@@ -1,636 +1,526 @@
- * a `WriteSideEffectInstruction`.
+ * The reason for this predicate is a bit annoying:
+ * We cannot mark a `PointerArithmeticInstruction` that computes an offset based on some SSA
+ * variable `x` as a use of `x` since this creates taint-flow in the following example:
+ * ```c
+ * int x = array[source]
+ * sink(*array)
+ * ```
+ * This is because `source` would flow from the operand of `PointerArithmeticInstruction` to the
+ * result of the instruction, and into the `IndirectOperand` that represents the value of `*array`.
+ * Then, via use-use flow, flow will arrive at `*array` in `sink(*array)`.
+ *
+ * So this predicate recurses back along conversions and `PointerArithmeticInstruction`s to find the
+ * first use that has provides use-use flow, and uses that target as the target of the `nodeFrom`.
  */
-Instruction getDestinationAddress(Instruction instr) {
-  result =
-    [
-      instr.(StoreInstruction).getDestinationAddress(),
-      instr.(WriteSideEffectInstruction).getDestinationAddress()
-    ]
+private predicate adjustForPointerArith(Node nodeFrom, UseOrPhi use) {
+  nodeFrom = any(PostUpdateNode pun).getPreUpdateNode() and
+  exists(DefOrUse defOrUse, Node adjusted |
+    indirectConversionFlowStep*(adjusted, nodeFrom) and
+    nodeToDefOrUse(adjusted, defOrUse) and
+    adjacentDefRead(defOrUse, use)
+  )
 }
 
-/** Gets the source address of `instr` if it is an instruction that behaves like a `LoadInstruction`. */
-Instruction getSourceAddress(Instruction instr) { result = getSourceAddressOperand(instr).getDef() }
+/** Holds if there is def-use or use-use flow from `nodeFrom` to `nodeTo`. */
+predicate defUseFlow(Node nodeFrom, Node nodeTo) {
+  // `nodeFrom = any(PostUpdateNode pun).getPreUpdateNode()` is implied by adjustedForPointerArith.
+  exists(UseOrPhi use |
+    adjustForPointerArith(nodeFrom, use) and
+    useToNode(use, nodeTo)
+  )
+  or
+  not nodeFrom = any(PostUpdateNode pun).getPreUpdateNode() and
+  exists(DefOrUse defOrUse1, UseOrPhi use |
+    nodeToDefOrUse(nodeFrom, defOrUse1) and
+    adjacentDefRead(defOrUse1, use) and
+    useToNode(use, nodeTo)
+  )
+}
 
-/**
- * Gets the operand that represents the source address of `instr` if it is an
- * instruction that behaves like a `LoadInstruction`.
- */
-Operand getSourceAddressOperand(Instruction instr) {
-  result =
-    [
-      instr.(LoadInstruction).getSourceAddressOperand(),
-      instr.(ReadSideEffectInstruction).getArgumentOperand()
-    ]
+/** Holds if `nodeTo` receives flow from the phi node `nodeFrom`. */
+predicate fromPhiNode(SsaPhiNode nodeFrom, Node nodeTo) {
+  exists(UseOrPhi use, IRBlock bb2, int i2, PhiNode phi, SourceVariable v |
+    use.asDefOrUse().hasIndexInBlock(bb2, i2, v) and
+    phi = nodeFrom.getPhiNode() and
+    phi.getSourceVariable() = v and
+    adjacentDefRead(phi, _, -1, bb2, i2) and
+    useToNode(use, nodeTo)
+  )
+  or
+  exists(PhiNode phiFrom, PhiNode phiTo |
```

Comment:
Did you deliberately intend to not use `lastRefRedef` here? As is, there could be an intermediate read between `nodeFrom` and `nodeTo`.

Reply:
Well spotted! Initially, I used `phiHasInputFromBlock` (before learning about `lastRefRedef`) in `adjacentDefRead`, which led to missing flow, and I fixed it by using `lastRefRedef`. I imagine this call to `phiHasInputFromBlock` can also be replaced by `lastRefRedef`. Thanks!

Reply:
Fixed in b692aed874732e8c584dd761ec773cd826f73665. Indeed, `lastRefRedef` was the correct predicate to use. With that change, all of our uses of `phiHasInputFromBlock` disappeared 🎉.